### PR TITLE
Fix r2 get_region_name

### PR DESCRIPTION
**Pull request by kriscon-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/kriscon-db/mlflow/pull/10803?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10803/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10803
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Reverting the R2 usage of get_region_name to the old implementation because cloudflare's implementation of head_bucket does not comport with AWS's

### How is this PR tested?

- [ x] Existing unit/integration tests
- [x ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/artifact/optimized_s3_artifact_repo.py [hunk 1]
```diff
@@ -52,7 +53,10 @@ def __init__(
         self._addressing_style = addressing_style
         self._s3_endpoint_url = s3_endpoint_url
         self.bucket, self.bucket_path = self.parse_s3_compliant_uri(self.artifact_uri)
-        self._region_name = self._get_region_name()
+        if region_name is None:
```
Comment (smurching) [inline mlflow/store/artifact/optimized_s3_artifact_repo.py:56]: Dumb question, is passing `region_name` explicitly necessary or if we override `self.get_region_name` in the R2 repo will the correct implementation just get called here? Fine with passing explicitly if needed

**Review by smurching — COMMENTED** [review]
Looks good! One minor question + it'd be nice to test against a Marketplace model to make sure this works

File: mlflow/store/artifact/optimized_s3_artifact_repo.py (hunk 1 above)
Comment (kriscon-db) [reply mlflow/store/artifact/optimized_s3_artifact_repo.py:56]: It's currently a private method in OptimizedS3ArtifactRepository, so there's nothing to override.  Also, this call is happening in the super class's constructor, not w/in the R2ArtifactRepository, so this felt like the fastest route to solving it.

**Review by smurching — APPROVED** [review]
LGTM

**Pull request merged**